### PR TITLE
Allocation QoS [5063]

### PR DESCRIPTION
**Pull request by MiguelCompany**
This PR is contains #436, #462 and #469 plus some changes to hide the external dependency on foonathan/memory from public include files.

Please refer to the PRs above for details.

### CODE REVIEW

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/1470/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/1436/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/1447/)

File: src/cpp/rtps/writer/StatefulWriter.cpp [hunk 1]
```diff
@@ -354,32 +354,35 @@ void StatefulWriter::send_any_unsent_changes()
...
-                    remote_readers.push_back(remoteReader->guid());
-                    locatorLists.push_back(remoteReader->remote_locators());
+                    locator_selector_.enable(remoteReader->guid());
                     expectsInlineQos |= remoteReader->expects_inline_qos();
                 }
 
+                if (locator_selector_.state_has_changed())
+                {
+                    group.flush_and_reset();
+                    network.select_locators(locator_selector_);
```
Comment (richiware) [inline src/cpp/rtps/writer/StatefulWriter.cpp]: Suggestion: this code is repeated and may aim to commit errors. Maybe can be put in one single location.

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/1586/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/1550/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/1561/)

File: src/cpp/rtps/builtin/discovery/endpoint/EDPSimpleListeners.cpp [hunk 2]
```diff
@@ -60,45 +62,75 @@ void EDPSimplePUBListener::onNewCacheChangeAdded(RTPSReader* reader, const Cache
...
+
+                if (updating)
+                {
+                    if (data->m_qos.canQosBeUpdated(temp_writer_data_.m_qos))
+                    {
+                        data->update(&temp_writer_data_);
+                    }
+                    else
+                    {
+                        logWarning(RTPS_EDP, "Received incompatible update for ReaderQos. reader_guid = " << data->guid());
```
[the remark this answers is not in the document]
Comment (MiguelCompany) [reply src/cpp/rtps/builtin/discovery/endpoint/EDPSimpleListeners.cpp]: This is done inside EDP::pairing_writer_proxy_with_any_local_reader

File: src/cpp/rtps/writer/StatefulWriter.cpp (hunk 1 above)
Comment (richiware) [reply src/cpp/rtps/writer/StatefulWriter.cpp]: It will be useful to describe the usage of `flush_and_reset` using doxygen.

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/1591/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/1555/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/1566/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/1610/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/1574/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/1585/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/1611/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/1575/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/1586/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/1613/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/1577/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/1588/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/1620/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/1584/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/1595/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/1625/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/1589/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/1600/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/1659/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/1619/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/1631/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/1660/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/1620/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/1632/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/1668/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/1628/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/1640/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/1669/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/1629/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/1641/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/1671/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/1631/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/1643/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/1672/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/1632/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/1644/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/1673/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/1633/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/1645/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/1674/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/1634/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/1646/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/1678/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/1638/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/1650/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/1679/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/1639/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/1651/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/1680/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/1640/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/1652/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/1683/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/1643/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/1655/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/1686/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/1646/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/1658/)

File: src/cpp/rtps/reader/StatefulReader.cpp [hunk 3]
```diff
@@ -188,27 +241,27 @@ bool StatefulReader::matched_writer_lookup(
 
     if(returnedValue)
     {
-        logInfo(RTPS_READER,this->getGuid().entityId<<" FINDS writerProxy "<< writerGUID<<" from "<< matched_writers.size());
+        logInfo(RTPS_READER,this->getGuid().entityId<<" FINDS writerProxy "<< writerGUID<<" from "<< matched_writers_.size());
     }
     else
     {
-        logInfo(RTPS_READER,this->getGuid().entityId<<" NOT FINDS writerProxy "<< writerGUID<<" from "<< matched_writers.size());
+        logInfo(RTPS_READER,this->getGuid().entityId<<" NOT FINDS writerProxy "<< writerGUID<<" from "<< matched_writers_.size());
```
Comment (LuisGP) [inline src/cpp/rtps/reader/StatefulReader.cpp]: Line too long.

File: src/cpp/rtps/reader/StatefulReader.cpp [hunk 4]
```diff
@@ -517,15 +527,16 @@ bool StatefulReader::change_received(
         }
     }
 
-    std::unique_lock<std::recursive_mutex> writerProxyLock(*prox->getMutex());
+    // TODO (Miguel C): Refactor this inside WriterProxy
```
Comment (LuisGP) [inline src/cpp/rtps/reader/StatefulReader.cpp]: Is this TODO still pending? Could be done now?

File: src/cpp/rtps/reader/StatelessReader.cpp [hunk 5]
```diff
@@ -294,50 +308,45 @@ bool StatelessReader::processDataFragMsg(CacheChange_t *incomingChange, uint32_t
...
-            {
-                return true;
-            }
-        }
+        return true;
     }
 
-    return false;
+    return std::any_of(matched_writers_.begin(), matched_writers_.end(), 
+        [writerId](const RemoteWriterGuids_t& writer)
```
Comment (LuisGP) [inline src/cpp/rtps/reader/StatelessReader.cpp]: Capture by reference.

File: src/cpp/rtps/reader/WriterProxy.cpp [hunk 6]
```diff
@@ -198,281 +237,347 @@ bool WriterProxy::maybe_add_changes_from_writer_up_to(const SequenceNumber_t& se
...
 {
-    std::lock_guard<std::recursive_mutex> guard(*mp_mutex);
+    std::lock_guard<std::recursive_mutex> guard(mutex_);
 
     // Check if CacheChange_t was already and it was already removed from changesFromW container.
-    if(seqNum <= changesFromWLowMark_)
+    if(seq_num <= changes_from_writer_low_mark_)
     {
-        logInfo(RTPS_READER, "Change " << seqNum << " <= than max available sequence number " << changesFromWLowMark_);
+        logInfo(RTPS_READER, "Change " << seq_num << " <= than max available sequence number " << changes_from_writer_low_mark_);
```
Comment (LuisGP) [inline src/cpp/rtps/reader/WriterProxy.cpp]: Line too long.

File: src/cpp/rtps/reader/WriterProxy.cpp [hunk 7]
```diff
@@ -198,281 +237,347 @@ bool WriterProxy::maybe_add_changes_from_writer_up_to(const SequenceNumber_t& se
...
 {
     size_t returnedValue = 0;
-    std::lock_guard<std::recursive_mutex> guard(*mp_mutex);
+    std::lock_guard<std::recursive_mutex> guard(mutex_);
 
-    if(seqNum > changesFromWLowMark_)
+    if(seq_num > changes_from_writer_low_mark_)
     {
-        for(auto ch = m_changesFromW.begin(); ch != m_changesFromW.end() && ch->getSequenceNumber() < seqNum; ++ch)
+        for(ChangeIterator ch = changes_from_writer_.begin(); ch != changes_from_writer_.end() && ch->getSequenceNumber() < seq_num; ++ch)
```
Comment (LuisGP) [inline src/cpp/rtps/reader/WriterProxy.cpp]: Line too long

File: src/cpp/rtps/reader/WriterProxy.cpp [hunk 8]
```diff
@@ -198,281 +237,347 @@ bool WriterProxy::maybe_add_changes_from_writer_up_to(const SequenceNumber_t& se
...
+        else if (final_flag && !liveliness_flag)
+        {
+            if (are_there_missing_changes())
+            {
+                heartbeat_response_->restart_timer();
+            }
+        }
+
+        //FIXME: livelinessFlag
+        if (liveliness_flag)//TODOG && attributes_->m_qos.m_liveliness.kind == MANUAL_BY_TOPIC_LIVELINESS_QOS)
```
Comment (LuisGP) [inline src/cpp/rtps/reader/WriterProxy.cpp]: FIX and TODO still pending?

File: src/cpp/transport/TCPv4Transport.cpp [hunk 9]
```diff
@@ -331,138 +331,10 @@ bool TCPv4Transport::fillUnicastLocator(
...
-        }
-
-        if (!bValid)
-        {
-            logError(RTCP, "There isn't any valid TCP Address on the whitelist");
-        }
-    }
-
-    return unicastResult;
+    // TODO (Miguel C): WAN / LAN considerations
```
Comment (LuisGP) [inline src/cpp/transport/TCPv4Transport.cpp]: Mandatory TODO. WAN is currently working in develop.

File: src/cpp/xmlparser/XMLParser.cpp [hunk 10]
```diff
@@ -2637,7 +2638,13 @@ XMLP_ret XMLParser::fillDataNode(tinyxml2::XMLElement* p_profile, DataNode<Parti
     for (p_aux0 = p_element->FirstChildElement(); p_aux0 != NULL; p_aux0 = p_aux0->NextSiblingElement())
     {
         name = p_aux0->Name();
-        if (strcmp(name, DEF_UNI_LOC_LIST) == 0)
+        if (strcmp(name, ALLOCATION) == 0)
+        {
+            // allocation
+            if (XMLP_ret::XML_OK != getXMLParticipantAllocationAttributes(p_aux0, participant_node.get()->rtps.allocation, ident))
```
Comment (LuisGP) [inline src/cpp/xmlparser/XMLParser.cpp]: Line too long.

File: src/cpp/xmlparser/XMLParser.cpp [hunk 11]
```diff
@@ -2953,6 +2960,12 @@ XMLP_ret XMLParser::fillDataNode(tinyxml2::XMLElement* p_profile, DataNode<Subsc
                 return XMLP_ret::XML_ERROR;
             subscriber_node.get()->setEntityID(static_cast<uint8_t>(i));
         }
+        else if (strcmp(name, MATCHED_PUBLISHERS_ALLOCATION) == 0)
+        {
+            // matchedPublishersAllocation - containerAllocationConfigType
+            if (XMLP_ret::XML_OK != getXMLContainerAllocationConfig(p_aux0, subscriber_node.get()->matched_publisher_allocation, ident))
```
Comment (LuisGP) [inline src/cpp/xmlparser/XMLParser.cpp]: Line too long.

File: include/fastrtps/rtps/common/LocatorSelector.hpp [hunk 12]
```diff
@@ -0,0 +1,297 @@
...
+
+    /**
+     * Remove an entry from this selector.
+     *
+     * @param entry Pointer to the LocatorSelectorEntry to remove.
+     */
+    bool remove_entry(const GUID_t& guid)
+    {
+        return entries_.remove_if(
+            [guid](LocatorSelectorEntry* entry)
```
[the remark this answers is not in the document]
Comment (MiguelCompany) [reply include/fastrtps/rtps/common/LocatorSelector.hpp]: It is already a reference, isn't it?

[the remark this answers is not in the document]
Comment (MiguelCompany) [reply include/fastrtps/rtps/common/LocatorSelector.hpp]: Checked with godbolt and it does a copy... changing

File: src/cpp/rtps/reader/WriterProxy.cpp (hunk 8 above)
Comment (MiguelCompany) [reply src/cpp/rtps/reader/WriterProxy.cpp]: Yes. Pending for manual liveliness implementation

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/1689/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/1649/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/1661/)

File: src/cpp/transport/TCPv4Transport.cpp (hunk 9 above)
Comment (LuisGP) [reply src/cpp/transport/TCPv4Transport.cpp]: I verified WAN is working.

File: src/cpp/rtps/reader/WriterProxy.cpp [hunk 13]
```diff
@@ -198,281 +237,350 @@ bool WriterProxy::maybe_add_changes_from_writer_up_to(const SequenceNumber_t& se
...
 
     return returnedValue;
 }
 
-void WriterProxy::lost_changes_update(const SequenceNumber_t& seqNum)
+void WriterProxy::lost_changes_update(const SequenceNumber_t& seq_num)
 {
-    logInfo(RTPS_READER,m_att.guid.entityId<<": up to seqNum: "<<seqNum);
-    std::lock_guard<std::recursive_mutex> guard(*mp_mutex);
+    logInfo(RTPS_READER, attributes_.guid().entityId <<": up to seq_num: "<<seq_num);
```
Comment (LuisGP) [inline src/cpp/rtps/reader/WriterProxy.cpp]:
```suggestion
    logInfo(RTPS_READER, attributes_.guid().entityId << ": up to seq_num: " << seq_num);
```

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/1692/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/1651/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/1663/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/1693/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/1652/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/1664/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/1753/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/1712/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/1724/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/1756/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/1715/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/1727/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/1758/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/1717/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/1729/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/1759/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/1718/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/1730/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/1762/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/1721/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/1733/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/1792/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/1751/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/1763/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/1822/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/1781/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/1793/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/1877/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/1836/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/1848/)

**Comment by richiware** [comment]
Build status:
                                * Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/${linux_number}/)
                                * Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/${windows_number}/)
                                * Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/${mac_number}/)

**Comment by richiware** [comment]
Build status:
                                * Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/null/)
                                * Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/null/)
                                * Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/null/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/0/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/0/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/0/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/0/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/0/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/0/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/67/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/70/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/0/)

File: examples/C++/RTPSTest_as_socket/TestWriterSocket.cpp [hunk 14]
```diff
@@ -27,72 +27,79 @@
...
 }
 
 void TestWriterSocket::run(uint16_t nmsgs)
 {
-	for(int i = 0;i<nmsgs;++i )
-	{
-		CacheChange_t * ch = mp_writer->new_change([]() -> uint32_t { return 255; }, ALIVE);
+    for(int i = 0;i<nmsgs;++i )
+    {
+        CacheChange_t * ch = mp_writer->new_change([]() -> uint32_t { return 255; }, ALIVE);
```
Comment (LuisGP) [inline examples/C++/RTPSTest_as_socket/TestWriterSocket.cpp]: Reader is not receiving the data.

File: src/cpp/rtps/builtin/data/ParticipantProxyData.cpp [hunk 15]
```diff
@@ -36,63 +36,70 @@ namespace eprosima {
...
 #endif
-    isAlive(pdata.isAlive),
-    m_properties(pdata.m_properties),
-    m_userData(pdata.m_userData),
-    mp_leaseDurationTimer(nullptr)
+    , isAlive(pdata.isAlive)
+    , m_properties(pdata.m_properties)
+    , m_userData(pdata.m_userData)
+    , lease_duration_event(nullptr)
+    , should_check_lease_duration(false)
```
Comment (richiware) [inline src/cpp/rtps/builtin/data/ParticipantProxyData.cpp]: We have to study if `m_readers` and `m_writers` attributes have to be copied too.

File: src/cpp/rtps/builtin/data/ParticipantProxyData.cpp [hunk 16]
```diff
@@ -391,17 +397,15 @@ void ParticipantProxyData::clear()
...
     m_guid = pdata.m_guid;
     m_VendorId[0] = pdata.m_VendorId[0];
     m_VendorId[1] = pdata.m_VendorId[1];
     m_availableBuiltinEndpoints = pdata.m_availableBuiltinEndpoints;
-    m_metatrafficUnicastLocatorList = pdata.m_metatrafficUnicastLocatorList;
-    m_metatrafficMulticastLocatorList = pdata.m_metatrafficMulticastLocatorList;
-    m_defaultUnicastLocatorList = pdata.m_defaultUnicastLocatorList;
-    m_defaultMulticastLocatorList = pdata.m_defaultMulticastLocatorList;
+    metatraffic_locators = pdata.metatraffic_locators;
+    default_locators = pdata.default_locators;
```
Comment (richiware) [inline src/cpp/rtps/builtin/data/ParticipantProxyData.cpp]: We have to study if m_readers and m_writers attributes have to be copied too.

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/0/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/73/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/66/)

File: src/cpp/rtps/builtin/data/ParticipantProxyData.cpp (hunk 16 above)
Comment (MiguelCompany) [reply src/cpp/rtps/builtin/data/ParticipantProxyData.cpp]: This method is only called when a new participant is discovered. The destination of the copy will always be a new ParticipantProxyData or one from the pool, so there is no need for m_readers and m_writers to be copied

File: src/cpp/rtps/builtin/data/ParticipantProxyData.cpp (hunk 15 above)
Comment (MiguelCompany) [reply src/cpp/rtps/builtin/data/ParticipantProxyData.cpp]: This method is only called from SecurityManager when a new participant is discovered and the corresponding DiscoveredParticipantInfo struct is created. Only participant info is used, so there is no need to copy m_readers and m_writers

File: src/cpp/rtps/reader/StatefulReader.cpp [hunk 17]
```diff
@@ -43,70 +43,134 @@ using namespace eprosima::fastrtps::rtps;
 
 StatefulReader::~StatefulReader()
 {
-    logInfo(RTPS_READER,"StatefulReader destructor.";);
-    for(std::vector<WriterProxy*>::iterator it = matched_writers.begin();
-            it!=matched_writers.end();++it)
+    logInfo(RTPS_READER,"StatefulReader destructor.");
+    is_alive_ = false;
```
[the remark this answers is not in the document]
Comment (MiguelCompany) [reply src/cpp/rtps/reader/StatefulReader.cpp]: done on ac863fc

File: include/fastrtps/rtps/reader/StatefulReader.h [hunk 18]
```diff
@@ -16,24 +16,31 @@
...
+
 #ifndef DOXYGEN_SHOULD_SKIP_THIS_PUBLIC
 
 #include "RTPSReader.h"
+#include "../../utils/collections/ResourceLimitedVector.hpp"
+#include "../common/CDRMessage_t.h"
+#include "../messages/RTPSMessageGroup.h"
+
 #include <mutex>
+#include <atomic>
```
[the remark this answers is not in the document]
Comment (MiguelCompany) [reply include/fastrtps/rtps/reader/StatefulReader.h]: Removed on ac863fc

File: include/fastrtps/rtps/reader/StatefulReader.h [hunk 19]
```diff
@@ -45,48 +52,65 @@ class StatefulReader:public RTPSReader
...
+         * @return True if correct.
+         */
+        bool liveliness_expired(const GUID_t& writer_guid);
+
+        /**
+         * Manually asserts liveliness on a matched writer.
+         * @param writer_guid GUID of the writer proxy for which livelines is asserted.
+         * @return True if correct.
+         */
+        bool assert_liveliness(const GUID_t& writer_guid);
```
[the remark this answers is not in the document]
Comment (MiguelCompany) [reply include/fastrtps/rtps/reader/StatefulReader.h]: Removed from header on f55772e

File: src/cpp/rtps/builtin/data/ParticipantProxyData.cpp (hunk 15 above)
Comment (MiguelCompany) [reply src/cpp/rtps/builtin/data/ParticipantProxyData.cpp]: Added comments on commit 8500a0c

File: src/cpp/rtps/builtin/data/ParticipantProxyData.cpp (hunk 16 above)
Comment (MiguelCompany) [reply src/cpp/rtps/builtin/data/ParticipantProxyData.cpp]: Added comments on commit 8500a0c

File: include/fastrtps/rtps/builtin/discovery/endpoint/EDPSimple.h [hunk 20]
```diff
@@ -146,20 +182,13 @@ class EDPSimple : public EDP
...
+    ReaderProxyData temp_reader_proxy_data_;
+    WriterProxyData temp_writer_proxy_data_;
 };
 
-
-// Default configuration values for EDP entities.
- extern const Duration_t edp_heartbeat_period;
- extern const Duration_t edp_nack_response_delay;
- extern const Duration_t edp_nack_supression_duration;
- extern const Duration_t edp_heartbeat_response_delay;
```
Comment (MiguelBarro) [inline include/fastrtps/rtps/builtin/discovery/endpoint/EDPSimple.h]: Default EDP times declarations are moved from EDPSimple.h to EDPServer.h.
If access to these members is no longer direct (but through new virtual APIs) then it doesn't make any sense to declare them in any header. They should be declared in EDPSimple.cpp (base class of EDPServer) as static (not reachable out of the compilation unit).

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/78/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/81/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/0/)

File: include/fastrtps/rtps/builtin/discovery/endpoint/EDPSimple.h (hunk 20 above)
Comment (MiguelCompany) [reply include/fastrtps/rtps/builtin/discovery/endpoint/EDPSimple.h]: Done on 3855427

Comment (MiguelBarro) [reply include/fastrtps/rtps/builtin/discovery/endpoint/EDPSimple.h]: success!!!! 🎉

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/0/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/85/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/0/)

**Comment by raquelalvarezbanos** [comment]
Test failures seem unrelated.

**Pull request merged**